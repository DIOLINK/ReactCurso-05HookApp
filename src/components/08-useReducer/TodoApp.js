import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import TodoList from './TodoList';
import TodoForm from './TodoForm';

import './styles.css';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
  // return [
  //   {
  //     id: new Date().getTime(),
  //     description: 'Learn React',
  //     done: false,
  //   },
  // ];
};
const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = { type: 'REMOVE_TODO', payload: todoId };
    dispatch(action);
  };
  const handleToggle = (todoId) => {
    const action = { type: 'TOGGLE_TODO', payload: todoId };
    dispatch(action);
  };
  const handleAddTodo = (newTodo) => {
    dispatch({ type: 'ADD_TODO', payload: newTodo });
  };
  return (
    <div>
      <h1>
        TodoApp (<small>{todos.length}</small>)
      </h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <h4>Add TODO</h4>
          <hr />
          <TodoForm handleAddTodo={handleAddTodo} />
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
