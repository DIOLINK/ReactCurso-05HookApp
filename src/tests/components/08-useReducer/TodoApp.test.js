import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { dummyTodo } from '../../fixtures/dummyTodo';

describe('Prubas en <TodoApp />', () => {
  const wrapper = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn(() => {});
  it('Debe de mostrar componente por defecto', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('Debe de mostrar componente con datos', () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find('TodoForm').prop('handleAddTodo')(dummyTodo[0]);
      wrapper.find('TodoForm').prop('handleAddTodo')(dummyTodo[1]);
    });
    expect(wrapper.find('small').text().trim()).toBe(
      dummyTodo.length.toString()
    );
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });
  it('Debe de llamar el evento handleDelete', () => {
    wrapper.find('TodoForm').prop('handleAddTodo')(dummyTodo[0]);
    expect(wrapper.find('small').text().trim()).toBe('1');
    wrapper.find('TodoList').prop('handleDelete')(dummyTodo[0].id);
    expect(wrapper.find('small').text().trim()).toBe('0');
  });
});
