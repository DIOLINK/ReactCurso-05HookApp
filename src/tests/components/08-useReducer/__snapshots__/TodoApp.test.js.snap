// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prubas en <TodoApp /> Debe de mostrar componente por defecto 1`] = `
<div>
  <h1>
    TodoApp (
    <small>
      0
    </small>
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <h4>
        Add TODO
      </h4>
      <hr />
      <TodoForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
